electron/src/main.ts

```diff
@@ -13,7 +13,146 @@ See the License for the specific language governing permissions and
+    }
+  }
+  
+  resetRnn() {
+    this.cellState = getInitialWeights(this.models)
+    this.hiddenState = getInitialWeights(this.models)
+    if (this.lastSample != null) {
+      this.lastSample.dispose();
+    }
+    this.lastSample = tf.scalar(PerformanceRnn.PRIMER_IDX, 'int32');
+
+    // TODO replace this
+    // currentPianoTimeSec = piano.now();
+    // pianoStartTimestampMs = performance.now() - currentPianoTimeSec * 1000;
+
+    // TODO: necessary?
+    this.generateStep(this.currentLoopId);
+  }
+
+  async generateStep(loopId: number) {
+    
+    const { lstmKernel1, lstmKernel2, lstmKernel3, lstmBias1, lstmBias2, lstmBias3 } = this.models
+
+    const lstm1 = (data: tf.Tensor2D, c: tf.Tensor2D, h: tf.Tensor2D) =>
+        tf.basicLSTMCell(PerformanceRnn.forgetBias, lstmKernel1, lstmBias1, data, c, h);
+    const lstm2 = (data: tf.Tensor2D, c: tf.Tensor2D, h: tf.Tensor2D) =>
+        tf.basicLSTMCell(PerformanceRnn.forgetBias, lstmKernel2, lstmBias2, data, c, h);
+    const lstm3 = (data: tf.Tensor2D, c: tf.Tensor2D, h: tf.Tensor2D) =>
+        tf.basicLSTMCell(PerformanceRnn.forgetBias, lstmKernel3, lstmBias3, data, c, h);
+
+    const [c, h, outputs] = tf.tidy(() => {
+      // Generate some notes.
+      const innerOuts: tf.Scalar[] = [];
+      for (let i = 0; i < PerformanceRnn.STEPS_PER_GENERATE_CALL; i++) {
+        // Use last sampled output as the next input.
+        const eventInput = tf.oneHot(
+          this.lastSample.as1D(), EVENT_SIZE).as1D();
+        // Dispose the last sample from the previous generate call, 
+        // since we kept it.
+        if (i === 0) {
+          this.lastSample.dispose();
+        }
+
+        // TODO: get conditioning
+        // const conditioning = getConditioning();
+        const axis = 0;
+        const input = conditioning.concat(eventInput, axis).toFloat();
+        const output =
+            tf.multiRNNCell([lstm1, lstm2, lstm3], input.as2D(1, -1), this.cellState, this.hiddenState);
+        this.cellState.forEach(c => c.dispose());
+        this.hiddenState.forEach(h => h.dispose());
+        this.cellState = output[0];
+        this.hiddenState = output[1];
+
+        const outputH = this.hiddenState[2];
+        const logits = outputH.matMul(this.models.fcW).add(this.models.fcB);
+
+        const sampledOutput = tf.multinomial(logits.as1D(), 1).asScalar();
+
+        innerOuts.push(sampledOutput);
+        this.lastSample = sampledOutput;
+      }
+      return [this.cellState, this.hiddenState, innerOuts]
+    });
+
+    for (let i = 0; i < outputs.length; i++) {
+      this.playOutput(outputs[i].dataSync()[0]);
+    }
+
+
+    // TODO: warn here
+    // if (piano.now() - currentPianoTimeSec > MAX_GENERATION_LAG_SECONDS) {
+    //   console.warn(
+    //       `Generation is ${piano.now() - currentPianoTimeSec} seconds behind, ` +
+    //       `which is over ${MAX_NOTE_DURATION_SECONDS}. Resetting time!`);
+    //   currentPianoTimeSec = piano.now();
+    // }
```

Comment:
will create and add a nice logger for n4m (which can be viewed eigther on max console or node.debug)